policy: Add Port Range Support for Policies Part 2/3

nathanjsweet opened the pull request:
This PR prepares the policy engine for adding port ranges
by enabling the underlying userspace cache to calculate
insertion, deletion, and lookups with port ranges, as well
as adding unit tests to ensure that the logic works. It does
not add support for adding policy port ranges at the API
level that will be addressed in the final PR.

The Policy CRD is modified by this PR without
supporting port ranges at the policy repository level
(this will be added in the final PR). This has to be done 
because the "PortProtocol" struct is shared by both 
the CRD (aka the API level) and the L4Filter struct
(aka the cache level).


See commits for details.

nathanjsweet committed:
policy: Add Unit Tests for Port Ranges

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Fix Deny Insertion Logic

Deny insertion has, heretofore, handled duplicate
deny entries by only checking for wildcard port
protocol entries as a way to prevent duplicate
deny entries or by ignoring them completely as
duplicate deny entries did not cause any problems.
This does not work with the more
precise de-duplication logic required by port
ranges, because overlapping ranges is a new
dimension of de-duplication that has to be
enforced.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

tommyp1ckles commented on pkg/policy/mapstate.go:
```diff
@@ -284,6 +284,29 @@ func (e *MapStateEntry) HasDependent(key Key) bool {
+	}
+	if e == nil || bEntry == nil {
+		return false
+	}
```
you could return false early here if `len(e.owner) != len(bEntry.owner)`

tommyp1ckles reviewed (approved):
endpoint changes lgtm

kaworu reviewed (approved):
Hubble changes LGTM

learnitall reviewed (approved):
API changes LGTM!

doniacld commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -548,15 +548,15 @@ spec:
                                 an optional transport protocol
                               properties:
+                                endPort:
+                                  description: EndPort can be an L4 port number.
```
Can it be something else than a L4 port number?

nathanjsweet replied:
No. Good point.

doniacld reviewed (approved):
One nit question, otherwise LGTM! 🫰

nathanjsweet committed:
policy: Add InvertedPortMask to MapState Key

- Add PortMask to MapKey structure, to make "0" the default value.
- Use PortRangeToMaskedPorts when creating keys for toMapState in L4 policies.
- Update maps/policy to account for mask.
- Introduce range logic into mapstate.
- Fix minor IsSupersetOf test issues.
- Add range unit tests to distillery tests.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
